### PR TITLE
Feat: Add SPDX licenses suggestions

### PR DESCRIPTION
**Pull request by idillon-sfl**

### CODE REVIEW

File: server/src/connectionHandlers/onCompletion.ts [hunk 1]
```diff
@@ -454,3 +467,40 @@ function convertExtraSymbolsToCompletionItems (uri: string): CompletionItem[] {
   })
   return completionItems
 }
+
+export const getRangeOfTextToReplace = (document: TextDocument, position: Position): Range => {
```
Comment (idillon-sfl) [inline server/src/connectionHandlers/onCompletion.ts:471]: We might want to use this on other places in order to improve the precision of the text being replaced. I'm thinking for example of SRC_URI

File: server/src/completions/completion-item-data.d.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/* --------------------------------------------------------------------------------------------
```
Comment (deribaucourt) [inline server/src/completions/completion-item-data.d.ts:1]: I don't see license suggestions when the assigment is empty.
```
# Press Ctrl+Space in the quotes
LICENSE = ""
# Although this one works
LICENSE:append = " MIT BS"
```
image

Comment (deribaucourt) [reply server/src/completions/completion-item-data.d.ts:1]: Otherwise excelent for showing deprecation and documentation 
image

File: server/src/completions/spdx-licenses.ts [hunk 3]
```diff
@@ -37,7 +60,27 @@ export const licenseCompletionItems: CompletionItem[] = spdxLicenses.map((licens
...
+      source: spdxLicenseDescription,
+      payload: license
+    } satisfies CompletionItemData
   }
 ))
+
+export const getSpdxLicenseCompletionResolve = async (item: CompletionItem): Promise<CompletionItem> => {
+  try {
+    const license = item.data.payload as SpdxLicense
+    const detailsResponse = await fetch(license.detailsUrl)
```
Comment (deribaucourt) [inline server/src/completions/spdx-licenses.ts:82]: Is the database loaded every time completion is requested? We should load it at most once, lazy if possible (only when needed the first time)

File: server/src/completions/spdx-licenses.ts [hunk 4]
```diff
@@ -54,21 +54,27 @@ const spdxLicensesFileContent = fs.readFileSync(spdxLicensesPath, 'utf8')
...
-      description: spdxLicenseDescription
-    },
-    documentation: 'Loading...',
-    data: {
-      source: spdxLicenseDescription,
-      payload: license
-    } satisfies CompletionItemData
-  }
-))
+export const getLicenseCompletionItems = (rangeOfText: Range): CompletionItem[] => {
```
Comment (deribaucourt) [inline server/src/completions/spdx-licenses.ts:57]: Bonus: have documentation hover on licenses, possibly in a follow up PR.

File: server/src/completions/spdx-licenses.ts [hunk 5]
```diff
@@ -54,21 +54,27 @@ const spdxLicensesFileContent = fs.readFileSync(spdxLicensesPath, 'utf8')
...
+  return spdxLicenses.map((license) => (
```
Comment (deribaucourt) [inline server/src/completions/spdx-licenses.ts:58]: Since this is fully implemented on the server side, it should work with neoVim. Can you test it? Also to ensure it didn't break the npm packaging :stuck_out_tongue:

File: server/src/connectionHandlers/onCompletion.ts [hunk 6]
```diff
@@ -462,6 +468,34 @@ function convertExtraSymbolsToCompletionItems (uri: string): CompletionItem[] {
   return completionItems
 }
 
+export const getRangeOfTextToReplace = (document: TextDocument, position: Position): Range => {
```
Comment (deribaucourt) [inline server/src/connectionHandlers/onCompletion.ts:471]: You could specify what range it is to replace. I think it's a word right? So rename getWordRange. You don't even need to specify it's for replacement. This function doesn't know what the word range will be used for. That way we are more susceptible to reuse it in other contexts

File: scripts/update-ref.sh [hunk 7]
```diff
@@ -87,3 +87,24 @@ rm -rf vscode
...
+TMP_FILE="tmp.txt"
+tail -n +5 $FILE > $TMP_FILE
+
+echo "#!/bin/bash" > $FILE
+echo "" >> $FILE
+
+git clone --depth 1 --filter=blob:none --sparse https://github.com/spdx/license-list-data.git
+cd license-list-data
+git fetch --tags
+TMP_TAG=$(git tag --sort=-v:refname | head -n 1)
```
Comment (deribaucourt) [inline scripts/update-ref.sh:102]: Make sure you test this, by checking out an older revision and running the script locally

File: server/src/__tests__/completions.test.ts [hunk 8]
```diff
@@ -1259,4 +1259,60 @@ describe('On Completion', () => {
...
+
+    if (completionItem === undefined) {
+      throw new Error('completionItem is undefined')
+    }
+
+    const resultAfterResolve = await onCompletionResolveHandler(completionItem)
+    expect(resultAfterResolve).toEqual(
+      expect.objectContaining({
+        ...expectedResultBeforeResolve,
+        documentation: "Bison Exception\n\nAs a special exception, you may create a larger work that contains part or all of the Bison parser skeleton and distribute that work under terms of your choice, so long as that work isn't itself a parser generator using the skeleton or a modified version thereof as a parser skeleton. Alternatively, if you modify or redistribute the parser skeleton itself, you may (at your option) remove this special exception, which will cause the skeleton and the resulting Bison output files to be licensed under the GNU General Public License without this special exception.\n\nThis special exception was added by the Free Software Foundation in version 2.2 of Bison.\n\n"
```
Comment (deribaucourt) [inline server/src/__tests__/completions.test.ts:1314]: you can use expect.stringContaining to make this a bit more concize

File: server/src/__tests__/completions.test.ts [hunk 9]
```diff
@@ -1259,4 +1259,60 @@ describe('On Completion', () => {
...
+    expect(resultBeforeResolve).toEqual(
+      expect.arrayContaining([
+        expect.objectContaining(expectedResultBeforeResolve)
+      ])
+    )
+
+    const completionItem = resultBeforeResolve.find((item) => item.label === licenseTested)
+
+    if (completionItem === undefined) {
+      throw new Error('completionItem is undefined')
```
Comment (deribaucourt) [inline server/src/__tests__/completions.test.ts:1307]: Why not use expect.isDefined? It will appear as a propper test failure rather than register an uncaught exception

File: server/src/__tests__/fixtures/completion.bb [hunk 10]
```diff
@@ -29,9 +29,11 @@ do_foo() {
     $D
 }
 
+LICENSE = "GPL-2.with-bisn-e"
```
Comment (deribaucourt) [inline server/src/__tests__/fixtures/completion.bb:32]: You could add another test case with an empty quote, like shown in my example, it currently fails so it's a good thing to test in the future!

**Review by deribaucourt — CHANGES_REQUESTED** [review]

**Comment by deribaucourt** [comment]
Very promising, a few implementation remarks. Keep it up!

File: server/src/completions/spdx-licenses.ts (hunk 3 above)
Comment (idillon-sfl) [reply server/src/completions/spdx-licenses.ts:89]: done

File: server/src/completions/spdx-licenses.ts (hunk 4 above)
Comment (idillon-sfl) [reply server/src/completions/spdx-licenses.ts:58]: I had the same idea in mind!

Lets do it in a further PR

File: server/src/completions/spdx-licenses.ts (hunk 5 above)
Comment (idillon-sfl) [reply server/src/completions/spdx-licenses.ts:59]: That one gives me lot of trouble. Spent too much time today trying to launch the language server on neovim :/

File: server/src/connectionHandlers/onCompletion.ts (hunk 6 above)
Comment (idillon-sfl) [reply server/src/connectionHandlers/onCompletion.ts:471]: There's different definitions of what is a word. In this `onCompletion.ts` file, we already have `analyzer.wordAtPointFromTextPosition`, which would return the whole strings with its quotes.
For example, if the user has their cursor anywhere in the string on `LICENSE = "GPL-2.0-or-later & LGPL-2.1-or-later"`, it will return `"GPL-2.0-or-later & LGPL-2.1-or-later"`

Javascript regexes have an other definition of word boundary: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Word_boundary_assertion.
For example, in `"GPL-2.0-or-later & LGPL-2.1-or-later"`, the words are `GPL`, `2`, `0`, `or`, `later` and so on. It seems to be what happen by default, and this is not what we want. If you remove this function (with its TextEdit) and use completion while typing the word `later`, you'll end up with something like `GPL-2.0-or-GPL-2.0-or-later`.

For completion, I determined we want replace whatever is located between quotes and whitespaces around the cursor. Since I am not aware of any terminology for that sort of text, I thought it would make sense to call it "textToReplace".

File: scripts/update-ref.sh (hunk 7 above)
Comment (idillon-sfl) [reply scripts/update-ref.sh:102]: I did 👌

File: server/src/__tests__/completions.test.ts (hunk 8 above)
Comment (idillon-sfl) [reply server/src/__tests__/completions.test.ts]: done

File: server/src/__tests__/completions.test.ts (hunk 9 above)
Comment (idillon-sfl) [reply server/src/__tests__/completions.test.ts:1307]: Otherwise Typescript will complain `completionItem` is undefined when calling `onCompletionResolveHandler(completionItem)`.

The only alternative I can think of would be something like
``` typescript
expect(completionItem).toBeDefined()
if (completionItem === undefined) {
  return
}
```

which is equally ridiculous, but one line longer.

In all cases, this is somewhat a pointless check since the previous `expect(resultBeforeResolve)` already checked there is an item with that label. If that line is ever reached, I'd consider it an error in the test itself, not in the code that is tested

File: server/src/__tests__/fixtures/completion.bb (hunk 10 above)
Comment (idillon-sfl) [reply server/src/__tests__/fixtures/completion.bb:32]: Good catch.

Fixed and tested.

File: server/src/connectionHandlers/onCompletion.ts (hunk 6 above)
Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts:471]: Word isn't appropriate since these expression have several of them. Those expressions are delimited by quotes and whitespace. Their specificity is that they cannot contain whitespaces. Some ideas:
 - text-span-without-whitespace
 - expression-no-space

File: server/src/__tests__/completions.test.ts (hunk 9 above)
Comment (deribaucourt) [reply server/src/__tests__/completions.test.ts:1307]: Ok

File: server/src/connectionHandlers/onCompletion.ts (hunk 1 above)
Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts:471]: Agreed

File: server/src/completions/spdx-licenses.ts (hunk 3 above)
Comment (deribaucourt) [reply server/src/completions/spdx-licenses.ts:89]: Ok but it's not lazy.
```typescript
const spdxLicensesPath = path.join(__dirname, '../../resources/spdx-licenses.json')
const spdxLicensesFileContent = fs.readFileSync(spdxLicensesPath, 'utf8')
export const spdxLicensesCollection = JSON.parse(spdxLicensesFileContent) as SpdxLicenseCollection
```
We'd have faster startup if this was delayed and cached to the first call of `getLicenseCompletionItems()`

File: server/src/connectionHandlers/onCompletion.ts [hunk 11]
```diff
@@ -96,6 +99,16 @@ function getBitBakeCompletionItems (textDocumentPositionParams: TextDocumentPosi
       ]
     }
 
+    const variablesAllowedForLicenseCompletion = ['LICENSE']
+    const isVariableAllowedForLicenseCompletion = analyzer.isStringContentOfVariableAssignment(documentUri, wordPosition.line, wordPosition.character, variablesAllowedForLicenseCompletion)
+    if (isVariableAllowedForLicenseCompletion && recipeLocalFiles !== undefined && word !== null) {
+      const textDocument = analyzer.getAnalyzedDocument(documentUri)?.document
+      if (textDocument !== undefined) {
+        const rangeOfText = getRangeOfTextToReplace(textDocument, textDocumentPositionParams.position)
+        return getLicenseCompletionItems(rangeOfText)
```
Comment (deribaucourt) [inline server/src/connectionHandlers/onCompletion.ts:108]: You should also add the operators with properly labeled category (`&, `|`).

Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts:108]: Possibly another PR if you want

File: server/src/completions/spdx-licenses.ts (hunk 5 above)
Comment (deribaucourt) [reply server/src/completions/spdx-licenses.ts:59]: Currently not working on NVIM, there was actually a regression which I fix in: !tree-sitter-bitbake.wasm

I don't know why the licenses don't show with NVIM, the package .tar.gz looks good. It would be nice to investigate. Tell me what you couldn't get working with NVIM. The documentation should be updated if something is missing.

Comment (idillon-sfl) [reply server/src/completions/spdx-licenses.ts:84]: Done.

I am pointing out the use of node-cache in order to eventually release the memory. I wonder if there are other places we could consider using it.

File: server/src/connectionHandlers/onCompletion.ts (hunk 6 above)
Comment (idillon-sfl) [reply server/src/connectionHandlers/onCompletion.ts:471]: After "solving" many issues, I realized my version of neovim was not supporting language servers. I updated it and the BitBake language server seems to be launched now. However, I still fail to have completion or any of the extension's functionalities, and I see `"server_request: no handler found for"	"bitbake/getRecipeLocalFiles"` in the logs.

I tried to trigger completion with `ctrl+x`, `ctrl+p`, and many other random combinations.

My init.lua currently looks like this (I've been trying many things):
```lua
local vim = vim
local Plug = vim.fn['plug#']

vim.call('plug#begin')
Plug('neoclide/coc.nvim', {['branch'] = 'release'})
Plug('neovim/nvim-lspconfig')
vim.call('plug#end')

--vim.lsp.omnifunc()
vim.lsp.buf.hover()

vim.api.nvim_create_autocmd({ "BufEnter" }, {
  pattern = { "*.bb"},
  callback = function()
      print("BufEnter event triggered for bitbake file")
      vim.notify("Starting bitbake LSP", vim.log.levels.INFO)

      vim.lsp.start({
      name = "bitbake",
      cmd = { "language-server-bitbake", "--stdio" }
    })
  end,
})
```

`:checkhealth lsp` shows the following:
```
vim.lsp: require("vim.lsp.health").check()

- LSP log level : WARN
- Log path: /home/io/.local/state/nvim/lsp.log
- Log size: 38 KB

vim.lsp: Active Clients ~
- bitbake (id=1, root_dir=~/Downloads/nvim/v:null, attached_to=[1])

vim.lsp: File watcher ~
- file watching "(workspace/didChangeWatchedFiles)" disabled on all clients
```

Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts:471]:
```
"server_request: no handler found for"	"bitbake/getRecipeLocalFiles"
```
Means the server is sending a request to the client, which isn't there so this throws. There's a completion for local files in SRC_URI which must be what's throwing here. It probably cancels the request because we don't catch this early enough. However it looks like we do in server.ts:209. In that case it probably creates an error log, but we properly catch it. It must be something else.

Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts:471]: SO in our case onCompletion.ts:104, recipeLocalFiles is expected to be undefined. I don't think this should impact LICENSEs? We should remove this condition, and make required changes if any.
```typescript
    const isVariableAllowedForLicenseCompletion = analyzer.isStringContentOfVariableAssignment(documentUri, wordPosition.line, wordPosition.character, variablesAllowedForLicenseCompletion)
    if (isVariableAllowedForLicenseCompletion && recipeLocalFiles !== undefined && word !== null) {
```

Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts:471]: Your setup looks to properly start the bitbake language server now since it's dsiplaying a log. Currently, the only completion that works is new variable assignments on empty lines. Don't you have that?

File: server/src/connectionHandlers/onCompletion.ts (hunk 11 above)
Comment (idillon-sfl) [reply server/src/connectionHandlers/onCompletion.ts]: You mean stuff like
```
LICENSE = "GPL-2.0-or-later | BSD-2-Clause 0BSD"
```
?

So when the user is typing, `&` and `|` should also be included in the suggestions?

Also I see those a lot in BitBake code, but I don't see them in [BitBake documentation](https://docs.yoctoproject.org/bitbake/bitbake-user-manual/bitbake-user-manual-ref-variables.html#term-LICENSE) or [SPDX documentation](https://spdx.github.io/spdx-spec/v3.0/annexes/SPDX-license-expressions/). In SPDX documentation, I rather see `AND`, `OR` and `WITH`.

Comment (deribaucourt) [reply server/src/connectionHandlers/onCompletion.ts]: Yes in bitbake, the operators are `&` and `|`, maybe the SPDX ones are also supported?
It could be nice to have suggestion for the operators appearing at the top of the list.

Comment (idillon-sfl) [reply server/src/connectionHandlers/onCompletion.ts]: Done.

I searched in Poky for occurrences of the SPDX operators with the approximate following regex: `license .?=.* (and|AND|or|OR|with|WITH)`. Could not find anything.

**Review by deribaucourt — APPROVED** [review]

**Comment by idillon-sfl** [comment]
oh no, there was a conflict to solve. It needs to be approved again :/

**Review by deribaucourt — APPROVED** [review]

File: server/src/connectionHandlers/onCompletion.ts (hunk 6 above)
Comment (idillon-sfl) [reply server/src/connectionHandlers/onCompletion.ts]: oh. No, I don't even have that :/
